File: Chapter_02_OBSERVER/WeatherStation/datas/WeatherData.java
package Chapter_02_OBSERVER.WeatherStation.datas;

import java.util.ArrayList;

import Chapter_02_OBSERVER.WeatherStation.iterfaces.IObserver;
import Chapter_02_OBSERVER.WeatherStation.iterfaces.ISubject;

public class WeatherData implements ISubject {

    private ArrayList<IObserver> observers;
    private float temperature;
    private float humidity;
    private float pressure;

    public float getTemperature() {
        return temperature;
    }

    public float getHumidity() {
        return humidity;
    }

    public float getPressure() {
        return pressure;
    }

    public WeatherData() {
        observers = new ArrayList<IObserver>();
    }

    @Override
    public void registerObserver(IObserver o) {
        observers.add((IObserver) o);
    }

    @Override
    public void removeObserver(IObserver o) {
        int i = observers.indexOf(o);
        if (i >= 0) {
            observers.remove(i);
        }
    }

    @Override
    public void notifyObservers() {
        for (IObserver observer : observers) {
            observer.update(this);
        }
    }

    public void measurementsChanged() {
        notifyObservers();
    }

    public void setMeasurements(float temperature, float humidity, float pressure) {
        this.temperature = temperature;
        this.humidity = humidity;
        this.pressure = pressure;
        measurementsChanged();
    }

}

File: Chapter_02_OBSERVER/CSharpWeatherStation/data/WeatherData.cs
public struct Weather
{
    public double Temperature { get; }
    public double Humidity { get; }
    public double Pressure { get; }

    public Weather(double temperature, double humidity, double pressure)
    {
        Temperature = temperature;
        Humidity = humidity;
        Pressure = pressure;
    }
}
class WeatherData : IObservable<Weather>
{
    private Weather currentWeather;
    private List<IObserver<Weather>> observers;

    public WeatherData()
    {
        this.observers = new List<IObserver<Weather>>();
    }

    public IDisposable Subscribe(IObserver<Weather> observer)
    {
        if (!observers.Contains(observer))
            observers.Add(observer);
        observer.OnNext(currentWeather);
        return new Unsubscriber(observers, observer);
    }

    public void SetMeasurements(Weather weather)
    {
        currentWeather = weather;
        NotifyObservers();
    }

    private void NotifyObservers()
    {
        foreach (var observer in observers)
            observer.OnNext(currentWeather);
    }



    private class Unsubscriber : IDisposable
    {
        private List<IObserver<Weather>> _observers;
        private IObserver<Weather> _observer;

        public Unsubscriber(List<IObserver<Weather>> observers, IObserver<Weather> observer)
        {
            this._observers = observers;
            this._observer = observer;
        }

        public void Dispose()
        {
            if (_observer != null && _observers.Contains(_observer))
                _observers.Remove(_observer);
        }
    }
}
